### PR TITLE
Check for premature end of stream in Async

### PR DESCRIPTION
**Pull request by rozza**
JAVA-1881

[Review on Reviewable](https://reviewable.io/reviews/rozza/mongo-java-driver/96)

### CODE REVIEW

File: driver-core/src/test/unit/com/mongodb/connection/InternalStreamConnectionSpecification.groovy [hunk 1]
```diff
@@ -197,6 +198,37 @@ class InternalStreamConnectionSpecification extends Specification {
...
+            handler.completed(-1)
+        }
+        def connection = getOpenedConnection()
+        def latch = new CountDownLatch(1);
+
+        when:
+        connection.receiveMessageAsync(messageId1, new SingleResultCallback<ResponseBuffers>() {
+            @Override
+            void onResult(final ResponseBuffers result, final Throwable t) {
+                latch.countDown();
```
Comment (jyemin) [inline driver-core/src/test/unit/com/mongodb/connection/InternalStreamConnectionSpecification.groovy]: Can this also assert that 

    t instance of MongoSocketReadException 

?

**Comment by jyemin** [comment]
Did you check NettyStream to see if it suffers similarly?

**Comment by jyemin** [comment]
Initial review complete.

**Comment by rozza** [comment]
Removed the test as it wasnt testing the right thing - like sync I cant see how to test it.
NettyStream doesnt appear to suffer - but I cant find anything on how it signals end of stream

**Comment by rozza** [comment]
Manually tested thanks to Net Link Conditioner and reproduced the Busy Loop in `AsynchronousSocketChannelStream` - also updated to pass the error to the handler.

Netty implementation isn't impacted the closed connection is caught in the event loop and the callback handler is passed the error.

**Comment by jyemin** [comment]
Can you explain why the test was not testing the right thing?  I'm not seeing it.

**Comment by rozza** [comment]
The test was testing a mocked stream and was returning -1 for a bytebuffer - causing groovy to throw a runtime exception.  The issue is I couldn't get the `BasicCompletionHandler.completed(-1, null)` to trigger.

**Comment by jyemin** [comment]
ok, I see.  

lgtm

**Pull request closed without merging**